File: bin/stacked-monitor.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative '.common/script_base'

# Script to quickly setup stacked monitors for 4-monitor configuration
# Configuration: Two 1920x1080 monitors stacked on left, main monitor in center, portrait monitor on right
class StackedMonitor < ScriptBase
  SPACING = 20 # Pixels between monitors

  # Script metadata for standardized help text
  def script_emoji
    '📺'
  end

  def script_title
    'Stacked Monitor Setup'
  end

  def script_description
    'Configures a 4-monitor setup: two 1920x1080 monitors stacked on left side,
main 3200x1800 monitor in center, and 1800x3200 portrait monitor on right.'
  end

  def script_arguments
    ''
  end

  def show_examples
    puts 'Examples:'
    puts "  #{script_name}                    # Configure stacked monitors"
    puts "  #{script_name} --debug           # Show current setup and debug info"
    puts "  #{script_name} --dry-run         # Preview what would be configured"
  end

  def validate!
    unless System.command?('displayplacer')
      log_error('displayplacer is not installed')
      log_info('Install with: brew install jakehilborn/jakehilborn/displayplacer')
      exit(1)
    end
    log_debug('displayplacer found')
    super
  end

  def run
    log_banner(script_title)

    displays_info = get_display_info
    show_current_setup(displays_info) if debug_mode?

    validate_display_count(displays_info)
    main_monitor, left_bottom, left_top, portrait_monitor = identify_displays(displays_info)
    configure_positions(main_monitor, left_bottom, left_top, portrait_monitor)

    command = build_command([main_monitor, left_bottom, left_top, portrait_monitor])

    if dry_run?
      show_dry_run(command, main_monitor, left_bottom, left_top, portrait_monitor)
    else
      execute_monitor_setup(command, main_monitor, left_bottom, left_top, portrait_monitor)
    end

    show_completion(script_title)
  end

  private

  def debug_mode?
    @options[:debug] || ENV['DEBUG'] == '1'
  end

  def validate_display_dependency
    return if System.command?('displayplacer')

    log_error('displayplacer is not installed')
    log_info('Install with: brew install jakehilborn/jakehilborn/displayplacer')
    exit(1)
  end

  def get_display_info
    log_debug('Getting display information...')

    display_list = execute_cmd('displayplacer list', description: 'Getting display information')
    unless display_list
      log_error('Failed to get display list')
      exit(1)
    end

    displays = display_list.split("\n\n").select { |block| block.include?('Persistent screen id:') }
    displays.map { |display| parse_display_info(display) }
  end

  def parse_display_info(display)
    type_match = display.match(/Type: (.+)/)
    type_str = type_match ? type_match[1] : 'Unknown'
    size_match = type_str.match(/(\d+)\s*inch/)
    size_inches = size_match ? size_match[1].to_i : 0

    {
      persistent_id: display.match(/Persistent screen id: (.+)/)&.[](1),
      contextual_id: display.match(/Contextual screen id: (.+)/)&.[](1),
      main: display.include?(' - main display'),
      type: type_str,
      resolution: display.match(/Resolution: (.+)/)&.[](1),
      hertz: display.match(/Hertz: (.+)/)&.[](1),
      color_depth: display.match(/Color Depth: (.+)/)&.[](1),
      scaling: display.match(/Scaling: (.+)/)&.[](1),
      origin: display.match(/Origin: \(([^)]+)\)/)&.[](1),
      rotation: display.match(/Rotation: (.+)/)&.[](1),
      size_inches: size_inches
    }
  end

  def validate_display_count(displays_info)
    return if displays_info.length == 4

    log_error("Found #{displays_info.length} displays, expected 4")
    log_info('Available displays:')
    displays_info.each_with_index do |display, i|
      puts "  #{i + 1}. #{display[:type]} - #{display[:resolution]}"
    end
    exit(1)
  end

  def identify_displays(displays_info)
    # Find displays by resolution patterns
    left_monitors = displays_info.select { |d| d[:resolution] == '1920x1080' }
    main_monitor = displays_info.find { |d| d[:resolution] == '3200x1800' }
    portrait_monitor = displays_info.find { |d| d[:resolution] == '1800x3200' }

    validate_monitor_config(left_monitors, main_monitor, portrait_monitor, displays_info)

    # Return all four monitors
    left_bottom = left_monitors[0]  # Bottom left monitor
    left_top = left_monitors[1]     # Top left monitor

    log_debug("Main (center): #{main_monitor[:type]}")
    log_debug("Portrait (right): #{portrait_monitor[:type]}")
    log_debug("Left bottom: #{left_bottom[:type]}")
    log_debug("Left top: #{left_top[:type]}")

    [main_monitor, left_bottom, left_top, portrait_monitor]
  end

  def validate_monitor_config(left_monitors, main_monitor, portrait_monitor, displays_info)
    errors = []

    errors << "Expected 2 monitors with 1920x1080 resolution, found #{left_monitors.length}" if left_monitors.length != 2
    errors << "Could not find main monitor (3200x1800)" if main_monitor.nil?
    errors << "Could not find portrait monitor (1800x3200)" if portrait_monitor.nil?

    return if errors.empty?

    errors.each { |error| log_error(error) }
    show_available_displays(displays_info)
    exit(1)
  end

  def show_available_displays(displays_info)
    log_info('Available displays:')
    displays_info.each_with_index do |display, i|
      puts "  #{i + 1}. #{display[:type]} - #{display[:resolution]} (#{display[:size_inches]}\")"
    end
  end

  def show_current_setup(displays_info)
    puts "\n🔍 CURRENT MONITOR SETUP DEBUG INFO"
    puts '=' * 50

    displays_info.each_with_index do |display, i|
      puts "\nDisplay #{i + 1}:"
      puts "  Type: #{display[:type]}"
      puts "  Size: #{display[:size_inches]}\" #{display[:size_inches] == 16 ? '(16-inch monitor)' : '(primary candidate)'}"
      puts "  Resolution: #{display[:resolution]}"
      puts "  Current Position: #{display[:origin] || 'Unknown'}"
      puts "  Persistent ID: #{display[:persistent_id]}"
      puts "  Main Display: #{display[:main] ? '✅ Yes' : '❌ No'}"
      puts "  Hertz: #{display[:hertz]}"
      puts "  Color Depth: #{display[:color_depth]}"
      puts "  Scaling: #{display[:scaling]}"
      puts "  Rotation: #{display[:rotation]}"
    end

    # Show spatial arrangement
    puts "\n📍 SPATIAL ARRANGEMENT:"
    sorted_displays = displays_info.sort_by do |d|
      [d[:origin]&.split(',')&.first&.to_i || 0, d[:origin]&.split(',')&.last&.to_i || 0]
    end

    sorted_displays.each do |display|
      x, y = display[:origin]&.split(',')&.map(&:to_i) || [0, 0]
      width, height = parse_resolution(display[:resolution])
      puts "  #{display[:type]} (#{display[:size_inches]}\"):"
      puts "    Position: (#{x}, #{y})"
      puts "    Size: #{width} x #{height}"
      puts "    Right edge: #{x + width}, Bottom edge: #{y + height}"
      puts "    #{display[:main] ? '👑 MAIN DISPLAY' : ''}"
    end

    # Analyze potential issues
    puts "\n⚠️  POTENTIAL ISSUES:"
    main_display = displays_info.find { |d| d[:main] }
    expected_main = displays_info.find { |d| d[:resolution] == '3200x1800' }
    if main_display && main_display[:resolution] != '3200x1800'
      puts "  🚨 Main display is #{main_display[:resolution]} (should be 3200x1800)"
    end

    # Check if displays are overlapping or have gaps
    puts "\n🔧 RECOMMENDED CONFIGURATION:"
    left_monitors = displays_info.select { |d| d[:resolution] == '1920x1080' }
    main_monitor = displays_info.find { |d| d[:resolution] == '3200x1800' }
    portrait_monitor = displays_info.find { |d| d[:resolution] == '1800x3200' }

    if left_monitors.length == 2 && main_monitor && portrait_monitor
      puts "  Main (#{main_monitor[:type]}): Should be at center (0,0)"
      puts '  Left Stack: Two 1920x1080 monitors stacked on left side'
      puts "  Portrait (#{portrait_monitor[:type]}): Should be on right side"
    else
      puts '  ❌ Unexpected display configuration'
    end

    puts "\n" + '=' * 50
  end

  def configure_positions(main_monitor, left_bottom, left_top, portrait_monitor)
    log_debug('Calculating monitor positions...')

    # Parse resolutions
    left_width, left_height = parse_resolution(left_bottom[:resolution])  # 1920x1080
    main_width, main_height = parse_resolution(main_monitor[:resolution]) # 3200x1800
    portrait_width, portrait_height = parse_resolution(portrait_monitor[:resolution]) # 1800x3200

    # Position main monitor at origin (center)
    main_monitor[:origin] = '0,0'

    # Position left stack to the left of main monitor
    left_stack_x = -(left_width + SPACING)

    # Calculate left stack positioning - monitor 2 (top) at 1/3 of main display height
    # Monitor 2 (left top) starts at 1/3 of main display height
    left_top_y = main_height / 3
    left_top[:origin] = "#{left_stack_x},#{left_top_y}"

    # Monitor 1 (left bottom) positioned below monitor 2
    left_bottom_y = left_top_y + left_height
    left_bottom[:origin] = "#{left_stack_x},#{left_bottom_y}"

    # Portrait monitor to the right of main monitor
    portrait_x = main_width + SPACING
    # Center portrait monitor vertically with main monitor
    main_center_y = main_height / 2
    portrait_y = main_center_y - (portrait_height / 2)
    portrait_monitor[:origin] = "#{portrait_x},#{portrait_y}"

    if debug_mode?
      log_debug("Main monitor: #{main_width}x#{main_height} at (0,0)")
      log_debug("Left stack: 2x #{left_width}x#{left_height} monitors at X: #{left_stack_x}")
      log_debug("Left top: Y #{left_top_y} to #{left_top_y + left_height} (at 1/3 of main height: #{main_height / 3})")
      log_debug("Left bottom: Y #{left_bottom_y} to #{left_bottom_y + left_height}")
      log_debug("Portrait: #{portrait_width}x#{portrait_height} at (#{portrait_x}, #{portrait_y})")
    end

    log_debug("Main position: #{main_monitor[:origin]}")
    log_debug("Left top position: #{left_top[:origin]}")
    log_debug("Left bottom position: #{left_bottom[:origin]}")
    log_debug("Portrait position: #{portrait_monitor[:origin]}")
  end

  def parse_resolution(resolution)
    width, height = resolution.split('x').map(&:to_i)
    [width, height]
  end

  def build_command(displays)
    command_parts = displays.map do |display|
      "id:#{display[:persistent_id]} " +
        "res:#{display[:resolution]} " +
        "hz:#{display[:hertz]} " +
        "color_depth:#{display[:color_depth]} " +
        "scaling:#{display[:scaling]} " +
        "origin:(#{display[:origin]}) " +
        "degree:#{display[:rotation]}"
    end

    'displayplacer ' + command_parts.map { |part| "\"#{part}\"" }.join(' ')
  end

  def show_configuration(main_monitor, left_bottom, left_top, portrait_monitor)
    puts "\n📺 Monitor Configuration:"
    puts "  Main Display (#{main_monitor[:type]}): Center position"
    puts "  Left Stack Bottom (#{left_bottom[:type]}): Left side, lower monitor"
    puts "  Left Stack Top (#{left_top[:type]}): Left side, upper monitor"
    puts "  Portrait Monitor (#{portrait_monitor[:type]}): Right side, vertical orientation"

    puts "\n📍 Calculated Positions:"
    puts "  Main: #{main_monitor[:origin]} (#{main_monitor[:resolution]})"
    puts "  Left Bottom: #{left_bottom[:origin]} (#{left_bottom[:resolution]})"
    puts "  Left Top: #{left_top[:origin]} (#{left_top[:resolution]})"
    puts "  Portrait: #{portrait_monitor[:origin]} (#{portrait_monitor[:resolution]})"
  end

  def show_dry_run(command, main_monitor, left_bottom, left_top, portrait_monitor)
    show_configuration(main_monitor, left_bottom, left_top, portrait_monitor)

    puts "\n🚀 Command to execute:"
    puts command
    puts "\nRun without --dry-run to execute automatically"
  end

  def execute_monitor_setup(command, main_monitor, left_bottom, left_top, portrait_monitor)
    show_configuration(main_monitor, left_bottom, left_top, portrait_monitor)

    log_progress('🔄 Executing monitor setup...')
    log_debug("Command: #{command}")

    success = execute_cmd?(command, description: 'Applying monitor configuration')
    if success
      log_success('Monitor arrangement completed successfully!')
    else
      log_error('Failed to execute displayplacer command')
      exit(1)
    end
  end

  def show_examples
    puts <<~EXAMPLES
      Examples:
        #{script_name}                    # Run interactive setup
        #{script_name} --dry-run          # Show configuration without applying
        #{script_name} --debug            # Show detailed setup analysis
        #{script_name} --dry-run --debug  # Show everything without applying
    EXAMPLES
  end
end

# Execute the script
StackedMonitor.execute if __FILE__ == $0

File: bin/bun-cli/src/scripts/macos/stacked-monitor.ts
import { Script } from '../../core/decorators/Script';
import { Script as BaseScript } from '../../core/base/Script';
import type { Context } from '../../core/types';
import { SystemService } from '../../core/services/SystemService';
import { execSilent } from '../../core/utils/shell';

interface DisplayInfo {
  persistentId: string;
  contextualId?: string;
  main: boolean;
  type: string;
  resolution: string;
  hertz: string;
  colorDepth: string;
  scaling: string;
  origin?: string;
  rotation: string;
  sizeInches: number;
}

interface MonitorPosition {
  monitor: DisplayInfo;
  x: number;
  y: number;
  width: number;
  height: number;
  right: number;
  bottom: number;
  rotation: number;
}

/**
 * Configure stacked external monitors
 *
 * Stacks two 1920x1080 16-inch external monitors vertically on the left or
 * right side of your main display. Shows ASCII visualization of layout.
 *
 * @example
 * stacked-monitor         # Stack on right side (default)
 * stacked-monitor left    # Stack on left side
 * stacked-monitor right   # Stack on right side
 */
@Script({
  emoji: '📺',
  tags: ['macos', 'display', 'monitors'],
  args: {
    direction: {
      type: 'string',
      position: 0,
      description: 'Stack direction: left or right (default: right)',
      enum: ['left', 'right'],
      default: 'right',
    },
    dryRun: {
      type: 'boolean',
      flag: '--dry-run',
      description: 'Preview configuration without applying',
    },
    debug: {
      type: 'boolean',
      flag: '--debug',
      description: 'Show detailed debug information',
    },
  },
})
export class StackedMonitorScript extends BaseScript {
  private readonly SPACING = 20;

  async validate(ctx: Context): Promise<void> {
    // Validate macOS
    if (!SystemService.isMacOS()) {
      throw new Error('This script requires macOS');
    }

    // Check for displayplacer
    if (!(await SystemService.commandExists('displayplacer'))) {
      throw new Error(
        'displayplacer is not installed. Install with: brew install jakehilborn/jakehilborn/displayplacer'
      );
    }
  }

  async run(ctx: Context): Promise<void> {
    const { direction, dryRun, debug } = ctx.args;

    this.logger.section('Stacked Monitor Setup');

    const displaysInfo = await this.getDisplayInfo();

    const externalMonitors = this.findExternalMonitors(displaysInfo);
    this.validateExternalMonitors(externalMonitors);

    // Show current configuration
    this.displayMonitorConfigBox(
      '🖥️  Current Monitor Configuration:',
      displaysInfo,
      true
    );
    this.displaySpatialLayout(displaysInfo, '📍 Current Monitor Layout');

    if (debug) {
      this.showCurrentSetup(displaysInfo);
    }

    // Configure external monitors
    this.configureExternalMonitors(externalMonitors, displaysInfo, direction);

    // Build command
    const command = await this.buildExternalCommand(displaysInfo, externalMonitors);

    if (!command) {
      this.logger.error('Failed to build displayplacer command');
      process.exit(1);
    }

    if (dryRun) {
      this.showDryRun(command, externalMonitors, displaysInfo, direction);
    } else {
      await this.executeMonitorSetup(command, externalMonitors, displaysInfo, direction);
    }

    this.logger.success('Stacked Monitor Setup Complete');
  }

  private async getDisplayInfo(): Promise<DisplayInfo[]> {
    try {
      const output = await execSilent('displayplacer list');
      const displayBlocks = output
        .split('\n\n')
        .filter((block) => block.includes('Persistent screen id:'));

      return displayBlocks.map((block) => this.parseDisplayInfo(block));
    } catch (error: any) {
      this.logger.error(`Failed to get display list: ${error.message}`);
      process.exit(1);
    }
  }

  private parseDisplayInfo(display: string): DisplayInfo {
    const typeMatch = display.match(/Type: (.+)/);
    const typeStr = typeMatch ? typeMatch[1] : 'Unknown';
    const sizeMatch = typeStr.match(/(\d+)\s*inch/);
    const sizeInches = sizeMatch ? parseInt(sizeMatch[1]) : 0;

    return {
      persistentId: display.match(/Persistent screen id: (.+)/)?.[1] || '',
      contextualId: display.match(/Contextual screen id: (.+)/)?.[1],
      main: display.includes(' - main display'),
      type: typeStr,
      resolution: display.match(/Resolution: (.+)/)?.[1] || '1920x1080',
      hertz: display.match(/Hertz: (.+)/)?.[1] || '60',
      colorDepth: display.match(/Color Depth: (.+)/)?.[1] || '8',
      scaling: display.match(/Scaling: (.+)/)?.[1] || 'on',
      origin: display.match(/Origin: \(([^)]+)\)/)?.[1],
      rotation: display.match(/Rotation: (.+)/)?.[1] || '0',
      sizeInches,
    };
  }

  private findExternalMonitors(displaysInfo: DisplayInfo[]): DisplayInfo[] {
    const externalMonitors = displaysInfo.filter(
      (d) => d.resolution === '1920x1080' && !d.type.includes('MacBook')
    );
    return externalMonitors;
  }

  private validateExternalMonitors(externalMonitors: DisplayInfo[]): void {
    if (externalMonitors.length !== 2) {
      this.logger.error(
        `Found ${externalMonitors.length} external 1920x1080 monitors, expected 2`
      );
      this.logger.info(
        'This script only works with two 16-inch external monitors (1920x1080).'
      );
      process.exit(1);
    }
  }

  private configureExternalMonitors(
    externalMonitors: DisplayInfo[],
    allDisplays: DisplayInfo[],
    direction: string
  ): void {
    const [width, height] = this.parseResolution(externalMonitors[0].resolution);

    // Find main display
    const mainDisplay = allDisplays.find((d) => d.main || d.type.includes('MacBook'));
    let referenceX = 0;
    let referenceY = 0;

    if (mainDisplay && mainDisplay.origin) {
      [referenceX, referenceY] = mainDisplay.origin.split(',').map(Number);
    }

    // Calculate stack position
    let stackX: number;
    if (direction === 'right') {
      stackX = referenceX + 2056 + this.SPACING; // MacBook width is 2056
    } else {
      stackX = referenceX - width;
    }

    // Stack monitors vertically
    const mainHeight = 1329; // MacBook height
    const bottomY = referenceY + Math.floor(mainHeight / 3);
    externalMonitors[0].origin = `${stackX},${bottomY}`;

    // Top monitor directly above bottom
    const topY = bottomY - height;
    externalMonitors[1].origin = `${stackX},${topY}`;
  }

  private parseResolution(resolution: string): [number, number] {
    const [width, height] = resolution.split('x').map(Number);
    return [width, height];
  }

  private async buildExternalCommand(
    allDisplays: DisplayInfo[],
    externalMonitors: DisplayInfo[]
  ): Promise<string | null> {
    try {
      const output = await execSilent('displayplacer list');
      const displayBlocks = output
        .split('\n\n')
        .filter((block) => block.includes('Persistent screen id:'));

      const commandParts: string[] = [];

      for (const block of displayBlocks) {
        const persistentId = block.match(/Persistent screen id: (.+)/)?.[1];
        const externalMonitor = externalMonitors.find(
          (ext) => ext.persistentId === persistentId
        );

        if (externalMonitor && externalMonitor.origin) {
          const config = this.buildDisplayConfigFromBlock(block, externalMonitor.origin);
          if (config) {
            commandParts.push(config);
          }
        }
      }

      if (commandParts.length === 0) {
        return null;
      }

      return 'displayplacer ' + commandParts.map((part) => `"${part}"`).join(' ');
    } catch (error: any) {
      this.logger.error(`Failed to build command: ${error.message}`);
      return null;
    }
  }

  private buildDisplayConfigFromBlock(
    displayBlock: string,
    newOrigin: string
  ): string | null {
    const persistentId = displayBlock.match(/Persistent screen id: (.+)/)?.[1];
    const resolution = displayBlock.match(/Resolution: (.+)/)?.[1];
    const hertz = displayBlock.match(/Hertz: (.+)/)?.[1];
    const colorDepth = displayBlock.match(/Color Depth: (.+)/)?.[1];
    const scaling = displayBlock.match(/Scaling: (.+)/)?.[1];
    const rotation = displayBlock.match(/Rotation: (.+)/)?.[1];

    if (!persistentId || !resolution || !hertz || !colorDepth || !scaling || !rotation) {
      return null;
    }

    return (
      `id:${persistentId} res:${resolution} hz:${hertz} color_depth:${colorDepth} ` +
      `scaling:${scaling} origin:(${newOrigin}) degree:${rotation}`
    );
  }

  private displayMonitorConfigBox(
    title: string,
    monitors: DisplayInfo[],
    showCurrentPositions: boolean = false
  ): void {
    console.log();
    console.log(title);
    console.log('┌' + '─'.repeat(78) + '┐');

    monitors.forEach((monitor, index) => {
      const label = (index + 1).toString();
      const name = monitor.type || 'Unknown Monitor';
      const resolution = monitor.resolution || 'Unknown';
      const rotation = parseInt(monitor.rotation || '0');

      let rotationIndicator = '';
      if (rotation === 90) rotationIndicator = ' ↻90°';
      else if (rotation === 180) rotationIndicator = ' ↻180°';
      else if (rotation === 270) rotationIndicator = ' ↻270°';

      const positionInfo =
        showCurrentPositions && monitor.origin ? ` at (${monitor.origin})` : '';

      const lineContent = `${label}. ${name} - ${resolution}${rotationIndicator}${positionInfo}`;
      const padding = Math.max(0, 76 - lineContent.length);

      console.log(`│ ${lineContent}${' '.repeat(padding)} │`);
    });

    console.log('└' + '─'.repeat(78) + '┘');
  }

  private displaySpatialLayout(monitors: DisplayInfo[], title: string): void {
    console.log();
    console.log(title);

    const monitorPositions: MonitorPosition[] = monitors.map((monitor) => {
      const [x, y] = monitor.origin ? monitor.origin.split(',').map(Number) : [0, 0];
      const [width, height] = this.parseResolution(monitor.resolution);
      const rotation = parseInt(monitor.rotation || '0');

      return {
        monitor,
        x,
        y,
        width,
        height,
        right: x + width,
        bottom: y + height,
        rotation,
      };
    });

    // Find layout bounds
    const minX = Math.min(...monitorPositions.map((m) => m.x));
    const maxX = Math.max(...monitorPositions.map((m) => m.right));
    const minY = Math.min(...monitorPositions.map((m) => m.y));
    const maxY = Math.max(...monitorPositions.map((m) => m.bottom));

    // Scale factor for ASCII representation
    const scaleX = Math.max((maxX - minX) / 60.0, 1);
    const scaleY = Math.max((maxY - minY) / 20.0, 1);

    // Create ASCII grid
    const gridWidth = 70;
    const gridHeight = 25;
    const grid: string[][] = Array.from({ length: gridHeight }, () =>
      Array(gridWidth).fill(' ')
    );

    // Draw each monitor
    monitorPositions.forEach((pos, index) => {
      const gridX = Math.round((pos.x - minX) / scaleX);
      const gridY = Math.round((pos.y - minY) / scaleY);
      const gridWidthM = Math.max(Math.round(pos.width / scaleX), 1);
      const gridHeightM = Math.max(Math.round(pos.height / scaleY), 1);

      const label = (index + 1).toString();
      let rotationChar = label;

      if (pos.rotation === 90 || pos.rotation === 270) rotationChar = '↻';
      else if (pos.rotation === 180) rotationChar = '↺';

      // Draw monitor box
      for (let dy = 0; dy < gridHeightM; dy++) {
        for (let dx = 0; dx < gridWidthM; dx++) {
          const gx = gridX + dx;
          const gy = gridY + dy;

          if (gx >= 0 && gx < gridWidth && gy >= 0 && gy < gridHeight) {
            if (dy === 0 || dy === gridHeightM - 1) {
              grid[gy][gx] = '─';
            } else if (dx === 0 || dx === gridWidthM - 1) {
              grid[gy][gx] = '│';
            } else if (
              dy === Math.floor(gridHeightM / 2) &&
              dx === Math.floor(gridWidthM / 2)
            ) {
              grid[gy][gx] = rotationChar;
            }
          }
        }
      }

      // Draw corners
      const corners: Array<[number, number, string]> = [
        [gridX, gridY, '┌'],
        [gridX + gridWidthM - 1, gridY, '┐'],
        [gridX, gridY + gridHeightM - 1, '└'],
        [gridX + gridWidthM - 1, gridY + gridHeightM - 1, '┘'],
      ];

      corners.forEach(([x, y, char]) => {
        if (x >= 0 && x < gridWidth && y >= 0 && y < gridHeight) {
          grid[y][x] = char;
        }
      });
    });

    // Print grid
    console.log('┌' + '─'.repeat(gridWidth) + '┐');
    grid.forEach((row) => {
      console.log('│' + row.join('') + '│');
    });
    console.log('└' + '─'.repeat(gridWidth) + '┘');

    // Print legend
    console.log('\nLegend:');
    monitors.forEach((monitor, index) => {
      const name = monitor.type || 'Unknown Monitor';
      const resolution = monitor.resolution || 'Unknown';
      const rotation = parseInt(monitor.rotation || '0');

      let rotationInfo = ' - Landscape';
      if (rotation === 90) rotationInfo = ' - Portrait 90° ↻';
      else if (rotation === 180) rotationInfo = ' - Inverted 180° ↺';
      else if (rotation === 270) rotationInfo = ' - Portrait 270° ↻';

      console.log(`  ${index + 1}. ${name} (${resolution})${rotationInfo}`);
    });
  }

  private showCurrentSetup(displaysInfo: DisplayInfo[]): void {
    console.log();
    console.log('🔍 DETAILED MONITOR DEBUG INFO');
    console.log('='.repeat(50));

    console.log();
    console.log('📊 Detailed Display Information:');
    displaysInfo.forEach((display, i) => {
      console.log();
      console.log(`Display ${i + 1}:`);
      console.log(`  Type: ${display.type}`);
      console.log(
        `  Size: ${display.sizeInches}" ${display.sizeInches === 16 ? '(16-inch monitor)' : ''}`
      );
      console.log(`  Resolution: ${display.resolution}`);
      console.log(`  Current Position: ${display.origin || 'Unknown'}`);
      console.log(`  Persistent ID: ${display.persistentId}`);
      console.log(`  Main Display: ${display.main ? '✅ Yes' : '❌ No'}`);
      console.log(`  Hertz: ${display.hertz}`);
      console.log(`  Color Depth: ${display.colorDepth}`);
      console.log(`  Scaling: ${display.scaling}`);
      console.log(`  Rotation: ${display.rotation}`);
    });

    console.log();
    console.log('='.repeat(50));
  }

  private showDryRun(
    command: string,
    externalMonitors: DisplayInfo[],
    allDisplays: DisplayInfo[],
    direction: string
  ): void {
    this.showExternalConfiguration(externalMonitors, direction);

    console.log();
    console.log('📍 Target Monitor Layout (After Configuration):');
    const targetDisplays = this.getTargetDisplayLayout(externalMonitors, allDisplays);
    this.displaySpatialLayout(targetDisplays, '📍 Complete Target Layout');

    console.log();
    console.log('🚀 Command to execute:');
    console.log(command);
    console.log();
    console.log('Run without --dry-run to execute automatically');
  }

  private showExternalConfiguration(
    externalMonitors: DisplayInfo[],
    direction: string
  ): void {
    console.log();
    console.log('📺 External Monitor Configuration:');

    console.log();
    console.log('📍 Configuration Description:');
    const stackSide = direction === 'right' ? 'Right' : 'Left';
    console.log(`  Stack Direction: ${stackSide} side`);
    console.log(`  Stack Top: Upper external monitor (${externalMonitors[1].resolution})`);
    console.log(
      `  Stack Bottom: Lower external monitor (${externalMonitors[0].resolution})`
    );
    console.log('  Other monitors: Will remain in current positions');
  }

  private getTargetDisplayLayout(
    externalMonitors: DisplayInfo[],
    allDisplays: DisplayInfo[]
  ): DisplayInfo[] {
    return allDisplays.map((display) => {
      const externalMonitor = externalMonitors.find(
        (ext) => ext.persistentId === display.persistentId
      );
      if (externalMonitor && externalMonitor.origin) {
        return { ...display, origin: externalMonitor.origin };
      }
      return display;
    });
  }

  private async executeMonitorSetup(
    command: string,
    externalMonitors: DisplayInfo[],
    allDisplays: DisplayInfo[],
    direction: string
  ): Promise<void> {
    this.showExternalConfiguration(externalMonitors, direction);

    this.logger.progress('🔄 Executing monitor setup...');

    try {
      await execSilent(command);
      this.logger.success('Monitor arrangement completed successfully!');

      console.log();
      console.log('🎯 Final Monitor Layout Applied:');
      const finalDisplays = this.getTargetDisplayLayout(externalMonitors, allDisplays);
      this.displaySpatialLayout(finalDisplays, '📍 Complete Final Layout');
    } catch (error: any) {
      this.logger.error(`Failed to execute displayplacer command: ${error.message}`);
      process.exit(1);
    }
  }
}
